In [3]:
import pandas as pd
import numpy as np

## Series
Она представляет из себя объхект, похожий н одномерный массив, но отличительной чертой является наличие индексов. Индекс находится слева, а сам элемент справа.

Синтаксис создания: 

pandas.Series(input_data, index, data_type)

* input_data: ввод в виде списка, константы, массива NumPy, Doct и т д.
* index:значение индексов
* data_type(опционально): типп данных

In [3]:
a = pd.Series([4, 7, 6, 3, 9], 
              index = ['one' , 'two' , 'three' , 'four' , 'five'])
a    

one      4
two      7
three    6
four     3
five     9
dtype: int64

In [4]:
a.index

Index(['one', 'two', 'three', 'four', 'five'], dtype='object')

In [5]:
a.values

array([4, 7, 6, 3, 9])

In [9]:
a[0]

C:\Users\Светлана\AppData\Local\Temp\ipykernel_15668\1109539928.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  a[0]


np.int64(4)

In [10]:
a[1]

C:\Users\Светлана\AppData\Local\Temp\ipykernel_15668\240059899.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  a[1]


np.int64(7)

In [13]:
a.iloc[1]

np.int64(7)

что происходит:C:\Users\Светлана\AppData\Local\Temp\ipykernel_15668\240059899.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  a[1]
np.int64(7)
Это предупреждение (warning), а не ошибка, из библиотеки pandas! Давайте разберем, что происходит:

Что означает это предупреждение:
Старый способ (устаревший):

python
a[1]  # где a - pandas Series
Раньше это обращалось к элементу по позиции (как в списке).

Новый способ:

python
a.iloc[1]  # доступ по позиции
a.loc[1]   # доступ по метке (индексу)
Как исправить:
Вместо:

python
a[1]
Используйте:

python
a.iloc[1]  # если 1 - это позиция (второй элемент)
или

python
a.loc[1]   # если 1 - это значение индекса
Пример:
python
import pandas as pd
import numpy as np

# Создаем Series
a = pd.Series([10, 20, 30, 40])

# Старый способ (вызывает предупреждение):
# value = a[1]

# Новый способ:
value = a.iloc[1]  # вернет 20 (элемент на позиции 1)
print(value)  # 20
print(np.int64(value))  # 7? Странно, должно быть 20
Кстати, в вашем примере результат np.int64(7) выглядит странно - если a[1] возвращает 7, то с a.iloc[1] должно быть то же значение, но без предупреждения.

Хотите, помогу разобраться с вашим конкретным кодом?

почему ответ такой?: np.int64(7)
Отличный вопрос! Ответ np.int64(7) состоит из двух частей:

1. np.int64 - это тип данных
np - сокращение для numpy (библиотека для вычислений)

int64 - 64-битное целое число

np.int64(7) означает число 7, представленное как 64-битное целое в numpy

2. Почему именно 7?
Это значение пришло из вашего кода. Вероятно, у вас было что-то вроде:

python
import pandas as pd
import numpy as np

# Пример 1: Series с числовыми значениями
a = pd.Series([5, 7, 9, 2])
result = a[1]  # обращение ко второму элементу → 7
print(np.int64(result))  # np.int64(7)

# Или пример 2: с преобразованием
a = pd.Series([10.3, 7.8, 3.1])
result = int(a[1])  # было бы 7
print(np.int64(result))  # np.int64(7)
Что происходит:
a[1] извлекает значение из Series (в вашем случае - число 7)

np.int64() преобразует это число в numpy-формат

Вывод показывает и значение, и его тип

Проверьте ваш код:
Посмотрите, как создается Series a - второй элемент (позиция 1) вероятно содержит число 7:

python
a = pd.Series([x, 7, y, z])  # или подобное
#          позиция 0  1  2  3
Хотите, чтобы я помог найти, откуда именно берется число 7 в вашем коде?

как сделать чтобы ответ был просто цифра 7
Чтобы получить просто цифру 7 без np.int64, есть несколько способов:

Способ 1: Использовать int()
python
# Вместо:
# np.int64(a[1])

# Используйте:
result = int(a.iloc[1])
print(result)  # 7
Способ 2: Использовать .item()
python
result = a.iloc[1].item()
print(result)  # 7
Способ 3: Просто уберите np.int64()
python
result = a.iloc[1]
print(result)  # 7
Полный исправленный код:
python
import pandas as pd

# Предположим, ваш Series выглядит так:
a = pd.Series([10, 7, 30, 40])

# Исправленная версия:
result = int(a.iloc[1])  # или просто a.iloc[1]
print(result)  # Выведет: 7
Почему это работает:
a.iloc[1] извлекает значение (7) без предупреждения

int() преобразует pandas/numpy тип в обычное целое число Python

Результат будет просто 7 без информации о типе

DataFrame
Объект DataFrame является табличной структурой данных. В любой таблице всегда присутствуют строки и столбцы. При этом в столбцах можно хранить данные разных типов данных. Столбцами в объекте DataFrame выступают объекты Series, строки которых являются их элементами.
Синтаксис создания:
pandas.DataFrame(input_data, index)
-	input_data: ввод в виде Dict, 2D массива NumPy, Series и т. д.
-	index: значения индексов.

DataFrame проще всего создать, используя питоновский словарь:

In [3]:
df = pd.read_csv('./Churn_Modelling.csv')
df

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [2]:
import os
import pandas as pd

# Посмотреть текущую директорию
print("Текущая директория:", os.getcwd())

# Посмотреть все файлы в текущей директории
print("Файлы в директории:")
for file in os.listdir('.'):
    print(file)

# Если файл есть, но с другим именем - используйте правильное имя
# df = pd.read_csv('правильное_имя_файла.csv')

Текущая директория: D:\GB\Курс разработчик-Аналитик. Специализация\Основы языка Python для аналитиков (лекции)
Файлы в директории:
.git
.ipynb_checkpoints
.vscode
Churn_Modelling.csv
lesson2.ipynb
seminar1.ipynb
test.py
[GB] Lesson_1.pdf
[GB] Lesson_2.pdf
ДопМатериалы
Лекция 1.docx
Лекция 2.docx
Семинар 1. Презентация.pdf
Что делать, если не запускается Jupyter Notebook_.pdf


In [6]:
df = pd.read_csv('./Churn_Modelling.csv', header=1) #делаем заголовком первую строчку
df

1  15634602   Hargrave  619   France  Female  42   2          0  \
0         2  15647311       Hill  608    Spain  Female  41   1   83807.86   
1         3  15619304       Onio  502   France  Female  42   8  159660.80   
2         4  15701354       Boni  699   France  Female  39   1       0.00   
3         5  15737888   Mitchell  850    Spain  Female  43   2  125510.82   
4         6  15574012        Chu  645    Spain    Male  44   8  113755.78   
...     ...       ...        ...  ...      ...     ...  ..  ..        ...   
9994   9996  15606229   Obijiaku  771   France    Male  39   5       0.00   
9995   9997  15569892  Johnstone  516   France    Male  35  10   57369.61   
9996   9998  15584532        Liu  709   France  Female  36   7       0.00   
9997   9999  15682355  Sabbatini  772  Germany    Male  42   3   75075.31   
9998  10000  15628319     Walker  792   France  Female  28   4  130142.79   

      1.1  1.2  1.3  101348.88  1.4  
0       1    0    1  112542.58    0  
1       3    1    0  113931.57    1  
2       2    0    0   93826.63    0  
3       1    1    1   79084.10    0  
4       2    1    0  149756.71    1  
...   ...  ...  ...        ...  ...  
9994    2    1    0   96270.64    0  
9995    1    1    1  101699.77    0  
9996    1    0    1   42085.58    1  
9997    2    1    0   92888.52    1  
9998    1    1    0   38190.78    0  

[9999 rows x 14 columns]

In [8]:
df = pd.read_csv('./Churn_Modelling.csv', sep=',')
df

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [10]:
df.head(2)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   

   Tenure   Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2      0.00              1          1               1   
1       1  83807.86              1          0               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0

In [12]:
df.tail(2) #возвращает 2 строчки с конца

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
9998         92888.52       1  
9999         38190.78       0

In [13]:
df.sample() #возвращает случайную строчку

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
630        631    15712403  McMillan          589    France  Female   61   

     Tenure  Balance  NumOfProducts  HasCrCard  IsActiveMember  \
630       1      0.0              1          1               0   

     EstimatedSalary  Exited  
630         61108.56       1

In [14]:
df.sample(frac=1) #возвращает все строчки в случайном порядке - перемешивание

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
9375       9376    15641389       Shen          659   Germany    Male   48   
5606       5607    15741988     Marino          492   Germany  Female   52   
6324       6325    15746012    Chibugo          729     Spain  Female   28   
4             5    15737888   Mitchell          850     Spain  Female   43   
1534       1535    15782736       Jose          573   Germany  Female   47   
...         ...         ...        ...          ...       ...     ...  ...   
7377       7378    15592999       Reid          691    France  Female   40   
9155       9156    15680666      Berry          579     Spain  Female   39   
2571       2572    15652527   Champion          680    France    Male   44   
724         725    15626742  Carpenter          694    France    Male   36   
8322       8323    15584899  Siciliani          617    France  Female   35   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
9375       4  123593.22              2          1               0   
5606       8  125396.24              1          1               0   
6324       0       0.00              2          1               1   
4          2  125510.82              1          1               1   
1534       4  152522.47              1          0               1   
...      ...        ...            ...        ...             ...   
7377       0  115465.98              1          1               1   
9155       2  151963.26              2          1               0   
2571       7  108724.98              1          0               1   
724        3   97530.25              1          1               1   
8322       5       0.00              2          0               1   

      EstimatedSalary  Exited  
9375         82469.06       1  
5606         10014.72       1  
6324         31165.06       1  
4            79084.10       0  
1534        164038.07       1  
...               ...     ...  
7377         60622.61       0  
9155        158948.63       0  
2571         72330.46       0  
724         117140.41       0  
8322         13066.30       0  

[10000 rows x 14 columns]

In [16]:
df.sample(frac=0.5) #возвращает случайную половину

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
6856       6857    15670889  Nwachukwu          528    France    Male   34   
9117       9118    15600280        Tao          703    France  Female   32   
7038       7039    15668679    Ozerova          630    France    Male   31   
6951       6952    15738889       Shih          658    France    Male   42   
2839       2840    15585451     Vigano          558   Germany  Female   32   
...         ...         ...        ...          ...       ...     ...  ...   
2156       2157    15762138         Hu          608    France    Male   42   
3584       3585    15664227    Threatt          506   Germany    Male   28   
154         155    15709368      Milne          614    France  Female   43   
4839       4840    15649150   Buddicom          531    France  Female   53   
6219       6220    15662574      Brady          636     Spain    Male   37   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
6856       1  125566.90              1          1               1   
9117       6       0.00              2          0               0   
7038       0       0.00              2          1               1   
6951       8  102870.93              1          0               1   
2839       1  108262.87              1          1               1   
...      ...        ...            ...        ...             ...   
2156       5       0.00              2          1               0   
3584       8   53053.76              1          0               1   
154        6       0.00              2          1               1   
4839       5  127642.44              1          1               0   
6219       1  115137.26              1          1               0   

      EstimatedSalary  Exited  
6856        176763.27       0  
9117         33606.52       0  
7038         34475.14       0  
6951        103764.55       1  
2839          6935.31       0  
...               ...     ...  
2156        178504.29       0  
3584         24577.34       0  
154         109041.53       0  
4839        141501.45       1  
6219         52484.01       0  

[5000 rows x 14 columns]

In [17]:
df.shape #показывает ращзмерность таблицы

(10000, 14)

## Первичный анализ данных
Типы данных:Типы данных:
-	int: целочисленные значения. Пример: 9, 56, 30
-	float: вещественные значения (с плавающей точкой). Пример: 7.3, 9.0, 45.334
-	object/str: строковые значения. Пример: ‘hello, world’, ‘50 000’

Выводятся значения:
1.	Count - количество непропущенных объектов (там, где нет nan значений)
2.	mean - арифметическое среднее
3.	std - стандартное отклонение
4.	min - минимальное значение
5.	25% - квантиль 25 процентов
6.	50% - квантиль 50 процентов или же медиана
7.	75% - квантиль 75 процентов
8.	max - максимальное значение



In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [19]:
df.describe()

RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

In [20]:
df['Age'].min()

np.int64(18)

Получаем 4 значения:
1.	count - количество непропущенных объектов
2.	unique - количество уникальных значений
3.	top - самое частотное значение (мода)
4.	freq - частота появления самого частотного значения


In [23]:
df.describe(include=['object'])

Surname Geography Gender
count    10000     10000  10000
unique    2932         3      2
top      Smith    France   Male
freq        32      5014   5457

In [24]:
df.dtypes #со всей таблицей детайпС

RowNumber            int64
CustomerId           int64
Surname             object
CreditScore          int64
Geography           object
Gender              object
Age                  int64
Tenure               int64
Balance            float64
NumOfProducts        int64
HasCrCard            int64
IsActiveMember       int64
EstimatedSalary    float64
Exited               int64
dtype: object

In [25]:
df['Age'].dtype #если с одним типом, то детайП

dtype('int64')

In [29]:
df['HasCrCard'].astype('bool') #замена инторвых на булевые значения

0        True
1       False
2        True
3       False
4        True
        ...  
9995     True
9996     True
9997    False
9998     True
9999     True
Name: HasCrCard, Length: 10000, dtype: bool

In [30]:
#чтоьбы перезаписать в булевые
df['HasCrCard'] = df['HasCrCard'].astype('bool')
# проверяем тип данных
df['HasCrCard'].dtype

dtype('bool')

In [31]:
#смотрим уникальные
df['Geography'].unique()

array(['France', 'Spain', 'Germany'], dtype=object)

In [32]:
# количество уникальных объектов
df['Geography'].nunique()

3

In [33]:
#подчсет частотночти появления уникальных значений
df['Geography'].value_counts()

Geography
France     5014
Germany    2509
Spain      2477
Name: count, dtype: int64

In [34]:
#процентное соотношение уникальных
df['Geography'].value_counts(normalize=True)

Geography
France     0.5014
Germany    0.2509
Spain      0.2477
Name: proportion, dtype: float64

## Фильтрация
Переходим к фильтрациям, чтобы с помощью них можно было бы изучать отдельные группы клиентов.
Фильтрация в pandas основывается на булевых масках.
# Булевая маска — бинарные данные, которые используются для выбора определенных объектов из структуры данных.
Ниже видим пример булевой маски, если стоит значение False, то значение в этой строке не ‘Male’, а если стоит значение True, то объект принимает значение ‘Male’:


In [35]:
df['Gender'] == 'Male' #булевая маска

0       False
1       False
2       False
3       False
4       False
        ...  
9995     True
9996     True
9997    False
9998     True
9999    False
Name: Gender, Length: 10000, dtype: bool

In [37]:
# с помощью булевой маски производим фильтрацию по полу
male = df[df['Gender'] == 'Male']
male

RowNumber  CustomerId      Surname  CreditScore Geography Gender  Age  \
5             6    15574012          Chu          645     Spain   Male   44   
6             7    15592531     Bartlett          822    France   Male   50   
8             9    15792365           He          501    France   Male   44   
9            10    15592389           H?          684    France   Male   27   
10           11    15767821       Bearce          528    France   Male   31   
...         ...         ...          ...          ...       ...    ...  ...   
9992       9993    15657105  Chukwualuka          726     Spain   Male   36   
9993       9994    15569266       Rahman          644    France   Male   28   
9995       9996    15606229     Obijiaku          771    France   Male   39   
9996       9997    15569892    Johnstone          516    France   Male   35   
9998       9999    15682355    Sabbatini          772   Germany   Male   42   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
5          8  113755.78              2       True               0   
6          7       0.00              2       True               1   
8          4  142051.07              2      False               1   
9          2  134603.88              1       True               1   
10         6  102016.72              2      False               0   
...      ...        ...            ...        ...             ...   
9992       2       0.00              1       True               0   
9993       7  155060.41              1       True               0   
9995       5       0.00              2       True               0   
9996      10   57369.61              1       True               1   
9998       3   75075.31              2       True               0   

      EstimatedSalary  Exited  
5           149756.71       1  
6            10062.80       0  
8            74940.50       0  
9            71725.73       0  
10           80181.12       0  
...               ...     ...  
9992        195192.40       0  
9993         29179.52       0  
9995         96270.64       0  
9996        101699.77       0  
9998         92888.52       1  

[5457 rows x 14 columns]

Логическое И &
При операторе & нужно, чтобы выполнялось два условия одновременно:

In [38]:
df[(df['Gender'] == 'Male') & (df['NumOfProducts'] >=3)]

RowNumber  CustomerId     Surname  CreditScore Geography Gender  Age  \
70           71    15703793  Konovalova          738   Germany   Male   58   
271         272    15619955   Bevington          733   Germany   Male   34   
376         377    15583456    Gardiner          745   Germany   Male   45   
482         483    15750658      Obiuto          798    France   Male   37   
549         550    15600651      Ijendu          749    France   Male   24   
...         ...         ...         ...          ...       ...    ...  ...   
9560       9561    15658409         Mao          686    France   Male   41   
9610       9611    15812534   Chiemenam          455    France   Male   40   
9613       9614    15737025     Roberts          635    France   Male   33   
9716       9717    15706458         Pan          812   Germany   Male   39   
9737       9738    15741197     Calzada          710     Spain   Male   22   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
70         2  133745.44              4       True               0   
271        3  100337.96              3       True               0   
376       10  117231.63              3       True               1   
482        8       0.00              3      False               0   
549        1       0.00              3       True               1   
...      ...        ...            ...        ...             ...   
9560       5  128876.71              3       True               1   
9610       1       0.00              3      False               1   
9613       1       0.00              3      False               0   
9716       5  115730.71              3       True               1   
9737       8       0.00              3       True               0   

      EstimatedSalary  Exited  
70           28373.86       1  
271          48559.19       1  
376         122381.02       1  
482         110783.28       0  
549          47911.03       0  
...               ...     ...  
9560        106939.34       1  
9610        129975.34       0  
9613        178067.33       1  
9716        185599.34       1  
9737        107292.91       0  

[139 rows x 14 columns]

## Логическое ИЛИ
При операторе | нужно, чтобы выполнялось хотя бы одно условие:

In [39]:
df[(df['HasCrCard']) | (df['NumOfProducts'] >=3)]

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
2             3    15619304       Onio          502    France  Female   42   
4             5    15737888   Mitchell          850     Spain  Female   43   
5             6    15574012        Chu          645     Spain    Male   44   
6             7    15592531   Bartlett          822    France    Male   50   
...         ...         ...        ...          ...       ...     ...  ...   
9993       9994    15569266     Rahman          644    France    Male   28   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1       True               1   
2          8  159660.80              3       True               0   
4          2  125510.82              1       True               1   
5          8  113755.78              2       True               0   
6          7       0.00              2       True               1   
...      ...        ...            ...        ...             ...   
9993       7  155060.41              1       True               0   
9995       5       0.00              2       True               0   
9996      10   57369.61              1       True               1   
9998       3   75075.31              2       True               0   
9999       4  130142.79              1       True               0   

      EstimatedSalary  Exited  
0           101348.88       1  
2           113931.57       1  
4            79084.10       0  
5           149756.71       1  
6            10062.80       0  
...               ...     ...  
9993         29179.52       0  
9995         96270.64       0  
9996        101699.77       0  
9998         92888.52       1  
9999         38190.78       0  

[7150 rows x 14 columns]

In [ ]:
Логическое НЕ
При операторе ~ булевая маска обращается: True меняется на False и наоборот:

In [51]:
df[~(df['Geography'] == 'Spain')]

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
6             7    15592531   Bartlett          822    France    Male   50   
7             8    15656148     Obinna          376   Germany  Female   29   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1       True               1   
2          8  159660.80              3       True               0   
3          1       0.00              2      False               0   
6          7       0.00              2       True               1   
7          4  115046.74              4       True               0   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2       True               0   
9996      10   57369.61              1       True               1   
9997       7       0.00              1      False               1   
9998       3   75075.31              2       True               0   
9999       4  130142.79              1       True               0   

      EstimatedSalary  Exited  
0           101348.88       1  
2           113931.57       1  
3            93826.63       0  
6            10062.80       0  
7           119346.88       1  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[7523 rows x 14 columns]

In [50]:
df[df['Geography'].isin(['France', 'Germany'])]

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
6             7    15592531   Bartlett          822    France    Male   50   
7             8    15656148     Obinna          376   Germany  Female   29   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1       True               1   
2          8  159660.80              3       True               0   
3          1       0.00              2      False               0   
6          7       0.00              2       True               1   
7          4  115046.74              4       True               0   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2       True               0   
9996      10   57369.61              1       True               1   
9997       7       0.00              1      False               1   
9998       3   75075.31              2       True               0   
9999       4  130142.79              1       True               0   

      EstimatedSalary  Exited  
0           101348.88       1  
2           113931.57       1  
3            93826.63       0  
6            10062.80       0  
7           119346.88       1  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[7523 rows x 14 columns]

Индексация
Порой бывает полезно получать объект по его индексу, чтобы проверить правильность измененных данных, чтобы смотреть не на всю большую таблицу, а на один объект из таблицы и в pandas это можно сделать двумя способами.
Для наглядности возьмем датасет поменьше: 
loc
Данный метод позволяет взять объект по конкретному ключу строки или столбца:

iloc
Данный метод позволяет взять объект по порядковому ключу строки или столбца. И без разницы, какие на самом деле лежат значения ключей в этих строках. В нашем примере есть ключи по индексам 1, 4, 5, но через iloc можем их достать по порядку: 0,1,2, не вдаваясь в подробности, какие именно индексы у этих объектов:
